src/indexing.jl

```diff
@@ -1,822 +1,312 @@
+# Default error messages to help users with new types and to avoid subsequent stack overflows
+getindex(a::StaticArray, i::Int) = error("getindex(::$typeof(a), ::Int) is not defined.")
+setindex!(a::StaticArray, value, i::Int) = error("setindex!(::$(typeof(a)), value, ::Int) is not defined.")
+
 ######################
-## Linear Indexing  ##
+## Scalar Indexing  ##
 ######################
 
-# What to do about @boundscheck and @inbounds? It's worse sometimes than @inline, for tuples...
-@generated function getindex{SA<:StaticArray, S}(a::SA, inds::NTuple{S,Integer})
-    newtype = similar_type(SA, Size(S))
-    exprs = [:(a[inds[$i]]) for i = 1:S]
-
-    return quote
-        $(Expr(:meta, :inline, :propagate_inbounds))
-        return $(Expr(:call, newtype, Expr(:tuple, exprs...)))
-    end
-end
-
-# Static Vector indexing into AbstractArrays
-@generated function getindex{T, I <: Integer}(
-        a::AbstractArray{T}, inds::StaticVector{I}
-    )
-    S = length(inds)
-    newtype = similar_type(inds, T, Size(S))
-    exprs = [:(a[inds[$i]]) for i = 1:S]
-    return quote
-        $(Expr(:meta, :inline, :propagate_inbounds))
-        return $(Expr(:call, newtype, Expr(:tuple, exprs...)))
-    end
-end
-# Convert to StaticArrays using tuples
-# TODO think about bounds checks here.
-@generated function getindex{T, I <: Integer}(
-        m::AbstractArray{T}, inds1::StaticVector{I}, i2::Integer
-    )
-    S = length(inds1)
-    newtype = similar_type(inds1, T, Size(S)) # drop singular dimension like in base
-    exprs = [:(m[inds1[$j], i2]) for j = 1:S]
-    return Expr(:call, newtype, Expr(:tuple, exprs...))
-end
-
-@generated function getindex{T, I <: Integer}(
-        m::AbstractArray{T}, i1::Integer, inds2::StaticVector{I}
-    )
-    S = length(inds2)
-    newtype = similar_type(inds2, T, Size(S))
-    exprs = [:(m[i1, inds2[$j]]) for j = 1:S]
-    return Expr(:call, newtype, Expr(:tuple, exprs...))
-end
+# Note: all indexing behavior defaults to dense, linear indexing
 
-@generated function getindex{I1 <: Integer,I2 <: Integer, T}(
-        m::AbstractArray{T}, inds1::StaticVector{I1}, inds2::StaticVector{I2}
-    )
-    S1, S2 = length(inds1), lengths(inds2)
-    exprs = [:(m[inds1[$j1], inds2[$j2]]) for j1 = 1:S1, j2 = 1:S2]
-    return Expr(:call, SMatrix{S1, S2, T}, Expr(:tuple, exprs...)) # I guess SMatrix should be fine?
+@propagate_inbounds function getindex(a::StaticArray, inds::Int...)
```

Comment:
Do you really want to support `inds::Int...`, or would it be better to support only `i::Vararg{Int, N}` where `N` matches the dimensionality of the array? If you do that, then you get all the goodness of dropped 1s, etc, for free.

Also, you may not need this at all for `IndexStyle(a)::IndexLinear`.